src/node_report.cc

```diff
@@ -0,0 +1,1115 @@
+
+  isolate->GetStackSample(state, samples, arraysize(samples), &info);
+  if (static_cast<size_t>(info.vm_state) < arraysize(v8_states)) {
+    out << "JavaScript VM state: " << v8_states[info.vm_state] << "\n\n";
+  } else {
+    out << "JavaScript VM state: <unknown>\n\n";
+  }
+  if (event == kSignal_UV) {
+    out << "Signal received when event loop idle, no stack trace available\n";
+    return;
+  }
+  Local<StackTrace> stack = StackTrace::CurrentStackTrace(isolate, 255,
+                              StackTrace::kDetailed);
+  if (stack.IsEmpty()) {
+    out << "\nNo stack trace available from StackTrace::CurrentStackTrace()\n";
+    return;
+  }
+  // Print the stack trace, adding in the
+  // pc values from GetStackSample() if available
+  for (int i = 0; i < stack->GetFrameCount(); i++) {
+    if (static_cast<size_t>(i) < info.frames_count) {
+      PrintStackFrame(out, isolate, stack->GetFrame(i), i, samples[i]);
+    } else {
+      PrintStackFrame(out, isolate, stack->GetFrame(i), i, nullptr);
+    }
+  }
+}
+
+/*******************************************************************************
+ * Function to print a JavaScript stack frame from a V8 StackFrame object
+ *
+ ******************************************************************************/
+static void PrintStackFrame(std::ostream& out, Isolate* isolate,
+                            Local<StackFrame> frame, int i, void* pc) {
+  String::Utf8Value fn_name_s(isolate, frame->GetFunctionName());
+  String::Utf8Value script_name(isolate, frame->GetScriptName());
+  const int line_number = frame->GetLineNumber();
+  const int column = frame->GetColumn();
+  char buf[64];
+
+  // First print the frame index and the instruction address
+  if (pc != nullptr) {
+#ifdef _WIN32
+    snprintf(buf, sizeof(buf), "%2d: [pc=0x%p] ", i, pc);
+#else
+    snprintf(buf, sizeof(buf), "%2d: [pc=%p] ", i, pc);
+#endif
+    out << buf;
+  }
+
+  // Now print the JavaScript function name and source information
+  if (frame->IsEval()) {
+    if (frame->GetScriptId() == Message::kNoScriptIdInfo) {
+      out << "at [eval]:" << line_number << ":" << column << "\n";
+    } else {
+      out << "at [eval] (" << *script_name << ":" << line_number << ":"
+          << column << ")\n";
+    }
+    return;
+  }
+
+  if (fn_name_s.length() == 0) {
+    out << *script_name << ":" << line_number << ":" << column << "\n";
+  } else {
+    if (frame->IsConstructor()) {
+      out << *fn_name_s << " [constructor] (" << *script_name << ":"
+          << line_number << ":" << column << ")\n";
+    } else {
+      out << *fn_name_s << " (" << *script_name << ":" << line_number << ":"
+          << column << ")\n";
+    }
+  }
+}
+
+
+#ifdef _WIN32
+/*******************************************************************************
+ * Function to print a native stack backtrace
+ *
+ ******************************************************************************/
+void PrintNativeStack(std::ostream& out) {
```

Comment:
It would be really good to integrate this code with the existing `NativeSymbolDebuggingContext`.

Reply:
will see to that. In your opinion is it fair doing it through a follow-up PR as opposed to here? given that it is an internal function and there is no external change?

Reply:
@addaleax - I have implemented this now by leveraging `NativeSymbolDebuggingContext`. My understanding is that this abstracts the platform differences underneath, so I haven't covered for any specific platforms, please let me know if that isn't the case.

```gdb
 1: [pc=0xab0032] nodereport::TriggerNodeReport(v8::Isolate*, node::Environment*, nodereport::DumpEvent, char const*, char const*, char*, v8::MaybeLocal<v8::Value>) [./node]
 2: [pc=0xab102a] nodereport::TriggerReport(v8::FunctionCallbackInfo<v8::Value> const&) [./node]
 3: [pc=0xc10b82]  [./node]
 4: [pc=0xc116f1] v8::internal::Builtin_HandleApiCall(int, v8::internal::Object**, v8::internal::Isolate*) [./node]
 5: [pc=0x1b1d73a4fc3d]
```

Reply:
subsequent work revealed that it indeed abstracts the platform differences, so all good here.